kad-dht/README.md

```diff
@@ -0,0 +1,288 @@
+
+The lookup for public key entries is identical to a standard entry lookup,
+except that a custom `Validator` strategy is applied. It checks that equality
+`SHA256(value) == peerID` stands when:
+
+1. Receiving a response from a `GET_VALUE` lookup.
+2. Storing a public key in the DHT via `PUT_VALUE`.
+
+The record is rejected if the validation fails.
+
+## Provider records
+
+Nodes must keep track of which nodes advertise that they provide a given key
+(CID). These provider advertisements should expire, by default, after 24 hours.
+These records are managed through the `ADD_PROVIDER` and `GET_PROVIDERS`
+messages.
+
+_WIP (jhiesey): explain what actually happens when `Provide()` is called._
+
+For performance reasons, a node may prune expired advertisements only
+periodically, e.g. every hour.
+
+## Node lookups
+
+_WIP (raulk)._
+
+## Bootstrap process
+The bootstrap process is responsible for keeping the routing table filled and
+healthy throughout time. It runs once on startup, then periodically with a
+configurable frequency (default: 5 minutes).
+
+On every run, we generate a random node ID and we look it up via the process
+defined in *Node lookups*. Peers encountered throughout the search are inserted
+in the routing table, as per usual business.
+
+This process is repeated as many times per run as configuration parameter
+`QueryCount` (default: 1). Every repetition is subject to a `QueryTimeout`
+(default: 10 seconds), which upon firing, aborts the run.
+
+## RPC messages
+
+_WIP (jheisey): consider just dumping a nicely formatted and simplified
+protobuf._
+
+See [protobuf
+definition](https://github.com/libp2p/go-libp2p-kad-dht/blob/master/pb/dht.proto)
+
+On any error, the entire stream is reset. This is probably not the behavior we
+want.
+
+* `FIND_NODE(key bytes) -> (nodes PeerInfo[])` Finds the `ncp` (default:6) nodes
+closest to `key` from the routing table and returns an array of `PeerInfo`s. If
+a node with id equal to `key` is found, returns only the `PeerInfo` for that
+node.
+* `GET_VALUE(key bytes) -> (record Record, closerPeers PeerInfo[])` If `key` is
+a public key (begins with `/pk/`) and the key is known, returns a `Record`
+containing that key. Otherwise, returns the `Record` for the given key (if in
+the datastore) and an array of `PeerInfo`s for closer peers.
+* `PUT_VALUE(key bytes, value Record) -> ()` Validates `value` and, if it is
+valid, stores it in the datastore.
+* `GET_PROVIDERS(key bytes) -> (providerPeers PeerInfo[], closerPeers
+PeerInfo[])` Verifies `key` is a valid CID. Returns `providerPeers` if in the
+providers cache, and an array of closer peers.
+* `ADD_PROVIDER(key, providerPeers PeerInfo[]) -> ()` Verifies `key` is a valid
+CID. For each provider `PeerInfo` that matches the sender's id and contains one
+or more multiaddrs, that provider info is added to the peerbook and the peer is
+added as a provider for the CID.
+* `PING() -> ()` Tests connectivity to destination node. Currently never sent.
```

Comment:
Shouldn't that be removed entirely?

Reply:
@jhiesey

Reply:
I'm inclined to leave this in the spec for now, as handlers are implemented in both JS and Go, and it may turn out to be useful soon as an end-to-end connection test. If we decide we don't want it then let's remove it from the code too.

Reply:
https://github.com/libp2p/specs/pull/108/commits/e1442fa3f109944c8eda8368c781ef5935572806 deprecates the `PING` message type with the following description:

> * `PING`: Deprecated message type replaced by the dedicated [ping
  protocol][ping]. Implementations may still handle incoming `PING` requests for
  backwards compatibility. Implementations must not actively send `PING`
  requests.

Please comment on this pull request in case you would like to not deprecate `PING` messages.

Also see https://github.com/libp2p/go-libp2p-kad-dht/issues/31

> If we decide we don't want it then let's remove it from the code too.

For backwards compatibility I suggest to keep the handling logic. As far as I can tell go-libp2p and rust-libp2p never actively send `PING` messages.